controllers/topology/internal/extensions/patches/inline/json_patch_generator.go

```diff
@@ -0,0 +1,293 @@
+		return nil, errors.Errorf("failed to calculate value: .valueFrom is set, but both .valueFrom.variable and .valueFrom.template are set")
+	}
+
+	// Return raw value.
+	if patch.Value != nil {
+		return patch.Value, nil
+	}
+
+	// Return variable.
+	if patch.ValueFrom.Variable != nil {
+		value, ok := variables[*patch.ValueFrom.Variable]
+		if !ok {
+			return nil, errors.Errorf("failed to calculate value: variable %q does not exist", *patch.ValueFrom.Variable)
+		}
+		return &value, nil
+	}
+
+	// Return rendered value template.
+	value, err := renderValueTemplate(*patch.ValueFrom.Template, variables)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to calculate value")
+	}
+	return value, nil
+}
+
+// renderValueTemplate renders a template with the given variables as data.
+func renderValueTemplate(valueTemplate string, variables map[string]apiextensionsv1.JSON) (*apiextensionsv1.JSON, error) {
+	// Parse the template.
+	tpl, err := template.New("tpl").Parse(valueTemplate)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse template: %q", valueTemplate)
+	}
+
+	// Convert the flat variables map in a nested map, so that variables can be
+	// consumed in templates like this: `{{ .builtin.cluster.name }}`
+	// NOTE: Variable values are also converted to their Go types as
+	// they cannot be directly consumed as byte arrays.
+	data, err := calculateTemplateData(variables)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to calculate template data")
+	}
+
+	// Render the template.
+	var buf bytes.Buffer
+	if err := tpl.Execute(&buf, data); err != nil {
+		return nil, errors.Wrapf(err, "failed to render template: %q", valueTemplate)
+	}
+
+	// Unmarshal the rendered template.
+	// NOTE: The YAML library is used for unmarshalling, to be able to handle YAML and JSON.
+	value := apiextensionsv1.JSON{}
+	if err := yaml.Unmarshal(buf.Bytes(), &value); err != nil {
+		return nil, errors.Wrapf(err, "failed to unmarshal rendered template: %q", buf.String())
+	}
+
+	return &value, nil
+}
+
+// calculateTemplateData calculates data for the template, by converting
+// a flat variables map to a nested map and converting the variables to
+// their Go types.
+// Example:
+// * Input:
+//   map[string]apiextensionsv1.JSON{
+//     "builtin.cluster.name": {Raw: []byte(`"cluster-name"`},
+//     "integerVariable": {Raw: []byte("4")},
+//     "numberVariable": {Raw: []byte("2.5")},
+//     "booleanVariable": {Raw: []byte("true")},
+//   }
+// * Output:
+//   map[string]interface{}{
+//     "builtin": map[string]interface{}{
+//       "cluster": map[string]interface{}{
+//         "name": <string>"cluster-name"
+//       }
+//     },
+//     "integerVariable": <float64>4,
+//     "numberVariable": <float64>2.5,
+//     "booleanVariable": <bool>true,
+//   }
+func calculateTemplateData(variables map[string]apiextensionsv1.JSON) (map[string]interface{}, error) {
+	res := map[string]interface{}{}
```

Comment:
```suggestion
	res := make(map[string]interface{}{}, len(variables))
```

Reply:
Depending on how the builtin variables are nested len(variables) is not the size of the top-level map, e.g.

variables:
  builtin.cluster.name
  builtin.cluster.namespace
len =>2 

map will have len = 1

Reply:
Fixed, as builtin variables are now working entirely differently it even is the exactly correct length :)